src/dotnet/ToolPackage/ObtainTransaction.cs

```diff
@@ -0,0 +1,255 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.DotNet.Tools;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.ProjectModel;
+using System.Transactions;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    internal class ObtainTransaction : IObtainTransaction
+    {
+        private string _packageId;
+        private string _packageVersion;
+        private FilePath? _nugetconfig;
+        private string _targetframework;
+        private string _source;
+
+        private Lazy<string> _bundledTargetFrameworkMoniker;
+        private Func<FilePath> _getTempProjectPath;
+        private IProjectRestorer _projectRestorer;
+        private DirectoryPath _toolsPath;
+        private DirectoryPath _offlineFeedPath;
+        private DirectoryPath _stageDirectory;
+
+        internal ObtainTransaction(
```

Comment:
Passing 9 params is horrible. I am thinking not to move all logic, and keep them in ToolPackageObtainer. Pass all the 4 method by closure. Since the whole point of ObtainTransaction  is to maintain the state of obtain

Reply:
But the will make this code looks like javascript

Reply:
I really like the result of passing closure, https://github.com/dotnet/cli/pull/8513/files#diff-a4cc784115b6c012508288eecabf25f5R39

There are much fewer changes